### DISCUSSION TITLE
What does a runner pod cache across consecutive jobs?

### DISCUSSION THREAD

**Discussion by alexklibisz** [actions/actions-runner-controller#2963] [Questions] [answered]
Hi, I have a couple of questions about the caching behavior of runners managed by a HorizontallRunnerAutoscaler: 

1. When a new runner pod starts, can it run multiple consecutive jobs? Or is a new runner started for every job? My understanding is that a single runner pod can run multiple consecutive (not concurrent) jobs, but I just want to make sure this is true.
2. Assuming a runner can run multiple consecutive jobs, what kind of caching is persisted across the jobs? E.g., if two consecutive jobs each use actions/checkout@v3, does actions/checkout@v3 get cloned from Github once, or twice?

TIA and thanks for the excellent project

**Comment by Link-** [comment]
This is not an FAQ. Moved to Questions.

**Comment by alexklibisz** [comment, marked as answer]
I think I've figured it out on my own:

> When a new runner pod starts, can it run multiple consecutive jobs? Or is a new runner started for every job? My understanding is that a single runner pod can run multiple consecutive (not concurrent) jobs, but I just want to make sure this is true.

It can be used for multiple consecutive jobs. Even with `ephemeral: true`, which is pretty confusing. This can be tested by: set up a runner deployment with exactly 1 runner; run a workflow with two jobs that each specify that runner; observe that the each job runs on the same runner.

> Assuming a runner can run multiple consecutive jobs, what kind of caching is persisted across the jobs? E.g., if two consecutive jobs each use actions/checkout@v3, does actions/checkout@v3 get cloned from Github once, or twice?

ARC doesn't implement any sort of native caching. We can do various forms of caching using volumes on the pod (to cache across jobs) and persistent volumes (to cache across pods). There's a good example here: https://github.com/actions/actions-runner-controller/issues/2726#issuecomment-1734202361
